### PR TITLE
Add OpenVINO plugin ConvolutionBackpropData op

### PR DESCRIPTION
**Pull request by tzerrell**
Adds the ConvolutionBackpropData op to the PlaidML OpenVINO plugin. Includes several parts:

1. Actually adding the op
2. Reformatting the tests to make the linter happy
3. Extending the oplib convolution to allow it to infer the result size for ConvDerivMode::DATA. Add a new parameter to the oplib conv to enable this (IMO we don't want it on by default; it's forced to make assumptions about the possible size as multiple input data sizes correspond to the same output size and I don't think we should make such assumptions without an explicit request)
4. Add some fancy template metaprogramming code to the Convolution Fluent API to allow multiple integer vector types to be used as inputs.
5. Advances the OpenVINO reference

Sorry that these are bundled in one PR, ~but no part of this is tested without the other parts~ turns out some of the edsl cc_tests do test the fluent API changes, I could split that off if it's important but I'd prefer not to.

### CODE REVIEW

File: plaidml/op/op.h [hunk 1]
```diff
@@ -156,25 +156,54 @@ class convolution {
...
-    data_dilations_ = data_dilations;
+  template <typename T, typename std::enable_if<std::is_integral<T>::value>::type* = nullptr>
+  convolution& dilations(const std::vector<T>& dilations) {
+    dilations_ = edsl::make_tuple(dilations);
     return *this;
   }
 
-  convolution& filter_shape(const std::vector<int>& filter_shape) {
-    filter_shape_ = filter_shape;
+  template <typename T, typename std::enable_if<std::is_integral<T>::value>::type* = nullptr>
```
Comment (flaub) [inline plaidml/op/op.h:180]: Cool!

**Comment by tzerrell** [comment]
Turns out we're getting build failures due to errors like
```
error: no matching function for call to 'plaidml::op::convolution::strides(llvm::ArrayRef<int>&)'
```
It looks like with the old version, the compiler was fine implicitly casting an `ArrayRef<int>` to a `vector<int>`, but in the templated format, it can't figure out that it can cast `ArrayRef<int>` to a `vector<T>` with `T` == `int`.

We can't just make a version of the function taking an `ArrayRef` because we don't want to force anyone including `op.h` to include LLVM (I presume). It would work to make anyone passing in an `llvm::ArrayRef` explicitly cast it to a `std::vector`.

@flaub do you have an opinion on whether it's better to back out the templated convolution builder functions or better to require explicit casting from `llvm::ArrayRef` to `std::vector` when building a convolution?

**Comment by flaub** [comment]
I thought there was a `to_vector` or something that ArrayRef or LLVM already implements.

**Comment by flaub** [comment]
Try: `ArrayRef<T>::vec()`

**Comment by flaub** [comment]
I don't see any ArrayRef in this code...

File: plaidml/op/lib/ops.cc [hunk 2]
```diff
@@ -816,30 +855,33 @@ Value convolution(const Value& value) {
...
-    throw std::runtime_error("Convolution op expects 17 arguments");
+  if (args.size() != 18) {
+    throw std::runtime_error("Convolution op expects 18 arguments");
   }
   auto I_or_O = args[0].as_tensor();  // O if deriv_mode is DATA, else I
   auto F_or_O = args[1].as_tensor();  // O if deriv_mode is FILTER, else F
-  auto strides = args[2].as_int_tuple();
-  auto dilations = args[3].as_int_tuple();
-  auto data_dilations = args[4].as_int_tuple();
+  auto strides = args[2].is_none() ? std::vector<int64_t>{} : args[2].as_int_tuple();
```
Comment (flaub) [inline plaidml/op/lib/ops.cc:867]: Let's make a `Value::as_int_tuple_or_empty()`

**Comment by tzerrell** [comment]
> I don't see any ArrayRef in this code...

There isn't an ArrayRef here. The problem is in https://github.com/plaidml/plaidml/blob/plaidml-v1/networks/oplib/resnet50.cc#L27 continuing through line 36. That code compiles without this PR and fails to compile with this PR, with an error that reads in part
```
ERROR: /home/tim/plaidml/networks/oplib/BUILD:7:1: C++ compilation of rule '//networks/oplib:oplib' failed (Exit 1)
networks/oplib/resnet50.cc: In function 'plaidml::edsl::Tensor networks::oplib::{anonymous}::block(const plaidml::edsl::Tensor&, llvm::ArrayRef<plaidml::edsl::Tensor>, llvm::ArrayRef<plaidml::edsl::Tensor>, llvm::ArrayRef<int>, bool, llvm::StringRef)':
networks/oplib/resnet50.cc:36:38: error: no matching function for call to 'plaidml::op::convolution::strides(llvm::ArrayRef<int>&)'
                      .strides(strides)
                                      ^
In file included from networks/oplib/resnet50.cc:9:
./plaidml/op/op.h:163:16: note: candidate: 'template<class T, typename std::enable_if<std::is_integral<_Tp>::value>::type* <anonymous> > plaidml::op::convolution& plaidml::op::convolution::strides(const std::vector<T>&)'
   convolution& strides(const std::vector<T>& strides) {
                ^~~~~~~
./plaidml/op/op.h:163:16: note:   template argument deduction/substitution failed:
networks/oplib/resnet50.cc:36:38: note:   'llvm::ArrayRef<int>' is not derived from 'const std::vector<T>'
                      .strides(strides)
                                      ^
```
ArrayRef does indeed have a conversion to std::vector: https://llvm.org/doxygen/classllvm_1_1ArrayRef.html#a745a0df36a633dcbc4acfd0bf9dffe17. I don't know why it was able to implicitly convert in the non-template form but unable to implicitly convert in the templated form; my knowledge of templates isn't very deep.

**Comment by tzerrell** [comment]
I found a way to minimize the changes needed to use this code with ArrayRefs. We do need to explicitly give the type when using an ArrayRef, but if the code I added in the latest commit to `networks/oplib/resnet50.cc` looks reasonable then that works.

**Comment by flaub** [comment]
LGTM, provided we figure out what the macos test failures are all about.

**Comment by tzerrell** [comment]
> LGTM, provided we figure out what the macos test failures are all about.

macos failures went away when I rebased on to a more recent `plaidml-v1`

**Pull request merged**